### DISCUSSION TITLE
Classic Controller usb hid gamepad

### DISCUSSION THREAD

**Discussion by Jimmdean-1** [jfurcean/CircuitPython_WiiChuck#11] [Q&A]
Greetings - enjoying your work here and looking to build my own usb adapters to use my older classic controllers as straight-usb gamepads in Dolphin, and probably nes/snes classic controllers in OpenEMU.
You have code examples published for making nunchuks into usb devices using your driver (and the Adafruit hardware), but do you have any for your classic controller driver?  The necessary USB HID descriptor seems pretty standard and is published elsewhere, but I haven’t been able to locate any code for the communication itself - I assume because very few people would choose to use a classic controller instead of something already natively usb-compatible.  Or that I’m late to the game and people who have already did this did it in Arduino long ago.  Being new to this in general (let alone Python itself) writing all that myself from scratch seems like a pretty daunting task.  
thanks a lot.

**Comment by jfurcean** [comment]
It should be pretty straight forward to combine the [CircuitPython Adafruit-HID library](https://github.com/adafruit/Adafruit_CircuitPython_HID/blob/main/examples/hid_gamepad.py) with  the [Classic Controller examples](https://github.com/jfurcean/CircuitPython_WiiChuck/tree/main/examples) from this library. I am happy to help if you have questions. What microcontroller are you using?

**Comment by Jimmdean-1** [comment]
QT PY RP2040 (w) QT breakout adapter

**Comment by jfurcean** [comment]
I created a gist for now that maps the classic controller to a USB HID game pad. https://gist.github.com/jfurcean/6d15effe779367c8d51dddab12ba4cb9. I tested it on Linux and it works as expected, but had some issues on Mac OSx so your mileage may vary. Let me know if you have any questions.

**Reply by Jimmdean-1** [reply]
Did your triggers work as expected?  I just noticed in your driver that the code and/or comments for the classic controller are reversed.  I have the original classic controller, so I know the triggers are analog potentiometers (0-31) with the button click at the end.  The “pro” model is a misnomer here - by the time it was released the potentiometers we’re gone and analog triggers haven’t been seen since in a Nintendo controller (boo).  I suspected maybe you had the later “pro” model (0 or 31).

**Comment by jfurcean** [comment]
I have a third party version (0 or 31). Are you saying that I just need to reverse the `Pro` and `non-Pro` comments on this line? 
https://github.com/jfurcean/CircuitPython_WiiChuck/blob/b407fcd34fcbaae8fc482d124382bb68efffc492/wiichuck/classic_controller.py#L106

**Reply by Jimmdean-1** [reply]
Hard to say for me right now - the triggers do not work for me here, but that is primarily because they are not being account for in the code or the descriptor.  This descriptor is very basic and needs some work for my controller, but that's not really my primary problem.  I have been able to get limited functionality in Windows - all buttons and one joy, but only in Windows, and not even in the gamepad tester.  In MacOS I get nothing, which is my OS of choice for this.
What exactly were your MacOS results?  Do you have any idea which portion of the code it is that MacOS is not responding to?  My guess would be the descriptor, but it's just a guess.  I just don't know the tools or how to get feedback from the OS on the MacOS side.

**Reply by Jimmdean-1** [reply]
make anything of this? (from console)

AppleUserUSBHostHIDDevice:0x100000cc1 Matching has vendor DeviceUsagePage : ff0c bundleIdentifier com.apple.AppleUserHIDDrivers ioclass AppleUserHIDEventService but transport and vendorID is missing

**Comment by jfurcean** [comment]
Yeah, I didn't account for the analog triggers in my code or the device descriptor because my device doesn't have them. On the MacOS side it looked like I had an additional 6 buttons and an additional joystick axis that weren't usable and threw off the results. This is the tool I tested with on both mac and linux - https://gamepad-tester.com/. I am not sure the best way of trouble shooting on the MacOS side either. I posted in the [Adafruit Discord](https://adafru.it/discord), but haven't heard anything. I will try some more testing tonight.

**Reply by jfurcean** [reply]
@Jimmdean-1 removing the `usb_hid.Device.MOUSE` and the `usb_hid.Device.CONSUMER_CONTROL` from `boot.py` fixed my issue on macOS.

**Reply by Jimmdean-1** [reply]
What is your version & architecture on your Mac?

I'm not having the same results.  It's completely possible the hardware in a generic vs genuine Classic Controller is different enough that we're just not going to be on the same page.

Do you happen to have NES Classic  or SNES Classic controllers?  I've haven't tried mine yet - I'm not sure what that would tell me.

**Reply by jfurcean** [reply]
Below is my setup. The board needs to be unplugged and plugged back in for the `boot.py` to run properly.

- macOS Monterey Version 12.3
- MacBook Pro (13-inch, M1, 2020) - Apple M1

I have wireless NES/SNES controllers that work with the WiiChuck library, the data on all of these are the same so the controller you use shouldn't matter. The USB descriptor would probably have to be updated depending on how you want the controls to work.

 Does this [example](https://github.com/jfurcean/CircuitPython_WiiChuck/blob/main/examples/classic_controller_simpletest.py) work on your board?

**Reply by Jimmdean-1** [reply]
oh - yeah.  The driver and your simpletests are fine - first thing I tried.  I’m sure this is just the descriptor.  I think a v1 classic controller is just a little too unique for the standard published descriptor.  I’ll have to test but I suspect the nes classic and snes classic controllers would work fine with just about any descriptor - no joysticks or analogs, everything is a button.

I was using my iMac though (Intel).  I’ll test on my M1 MBA in the next couple days.

**Reply by Jimmdean-1** [reply]
I think I made some progress - it partially works on my MBA, but this time I let the keyboard assistant run and identified it as a standard keyboard - on my iMac I had exited out of the keyboard assistant.  In hindsight I think I got the board blacklisted, so it wouldn't work at all.
On the MBA I just need to figure out the triggers, but also the Y-axis on my joysticks is inverted.  You didn't mention that happening to you - I suspect it's something I need to swap in code.py, and not boot.py?

**Reply by jfurcean** [reply]
That's good news. There is probably a way to remove it from the black list it, but I am not sure what that is. I am not sure if any of my axes were inverted, but you should be able to modify the inputs into `gp.move_joysticks` by multiplying the values by `-1`. Only do it for the axes that are inverted. 

```py
    gp.move_joysticks(

        x=int(map_range(joysticks.lx, 0, 63, -127, 127)) * -1,
        y=int(map_range(joysticks.ly, 0, 63, -127, 127)) *  -1,
        z=int(map_range(joysticks.rx, 0, 31, -127, 127)) *  -1,
        r_z=int(map_range(joysticks.ry, 0, 31, -127, 127)) * -1,
    )
```

**Reply by Jimmdean-1** [reply]
I'll give it a try.  I bypassed the blacklist simply by running the keyboard assistant manually.  it knew right away I was talking about the adafruit device.
additional note:  while the gamepad is now identified and working in gamepad tester, the OS does not seem to see it as a game controller - it won't show up in either Dolphin or OpenEmu.  Did you happen to try an emulator on your end with this code?

**Reply by jfurcean** [reply]
I haven't, but I will. It could be because of the `usb_hid.Device.KEYBOARD` still being used in the descriptor. You could try without that.

**Reply by jfurcean** [reply]
I updated all the files in the [gist](https://gist.github.com/jfurcean/6d15effe779367c8d51dddab12ba4cb9). These updates include the analog triggers (not sure if its correct, but it works). It showed up in EMU once I removed the `usb_hid.Device.KEYBOARD`, unplugged it, and plugged it back it. You will need to change the values for the trigger buttons to `0` in `code.py` if you want the analog triggers to work instead of using them as buttons.

**Reply by Jimmdean-1** [reply]
Your revised code and hid_gamepad are fine as-is.
I put my revised boot.py in the gist.  I had to tweak the analogs a little to make them act a little more like the axis of the joysticks, and make sure they had unique usage numbers - you had 2 with a 33.
Once I did that it wasn't pretty in gamepad-tester, but it was completely mappable in Dolphin with perfect functionality.

**Reply by jfurcean** [reply]
Thanks for catching that error and for testing this. I ended up putting together a [repo for this work](https://github.com/jfurcean/CircuitPython_WiiGamepads/) and included examples for the wii nunchuk, the wii classic controller, the nes controller, and the wii guitar hero controller.

**Reply by dglaude** [reply]
I don't find that repo: https://github.com/jfurcean/CircuitPython_WiiGamepads/ is it public?

I only noticed this very interesting conversation now... but I think this is great idea.
There are already product to get a retro joystick and turn it into USB, but here if we convert I2C Wii controller to USB and this is understood by emulator, it is really great (as I am collecting all the Wii accessories I can find).
Thank you very much both of you, this work need to be known.

PS: My attempt at making a custom USB descriptor were a total failure, also I had two SNES like USB controller, and they were not advertising the same way so I did not knew where to start.

**Reply by jfurcean** [reply]
Sorry @dglaude, it looks like I forgot to make it public. Just made the change.

**Comment by Jimmdean-1** [comment]
same here - and the negatives worked as well for the joys.  I’ll test out the rest this weekend - funny thing about the v1 classic controller is the triggers are analog to begin, then a digital button press (same as a Gamecube controller).  I don’t think the code will matter as even though they are shared the signals come from different components of the board.
